README.md
# Date Range Dropdown
[![Pub](https://img.shields.io/pub/v/date_range_dropdown?include_prereleases)](https://pub.dev/packages/date_range_dropdown)


Set predefined date ranges with dropdown.

## Demo
![](demogif.gif)

## Getting Started

### Installation

Add to `pubspec.yaml` in `dependencies`

```
date_range_dropdown: ^1.0.2
```

### Usage

```
import 'package:date_range_dropdown/date_range_dropdown.dart';

List<DatePickerModel> dates = List<DatePickerModel>();

dates.add(DatePickerModel.fromJson({
    "id": 'today',
    "title": "Today",
    "startDate": Jiffy().startOf(Units.DAY),
    "endDate": Jiffy().endOf(Units.DAY)
}));

dates.add(DatePickerModel.fromJson({
    "id": "custom",
    "title": "Custom",
    "startDate": DateTime.now(),
    "endDate": DateTime.now()
}));

DateRangeDropdown(
    dates: dates,
    onChanged: (DatePickerModel value) => setState(
        () {
        _selected = value.id;
        _startDate = value.startDate;
        _endDate = value.endDate;
        },
    ),
    callback: (DateTimeRange picked) => setState(
        () {
        _startDate = picked.start;
        _endDate = picked.end;
        },
    ),
)
```

### Theme Customization

```
import 'package:date_range_dropdown/date_range_dropdown.dart';

List<DatePickerModel> dates = List<DatePickerModel>();

dates.add(DatePickerModel.fromJson({
    "id": 'today',
    "title": "Today",
    "startDate": Jiffy().startOf(Units.DAY),
    "endDate": Jiffy().endOf(Units.DAY)
}));

dates.add(DatePickerModel.fromJson({
    "id": "custom",
    "title": "Custom",
    "startDate": DateTime.now(),
    "endDate": DateTime.now()
}));

DateRangeDropdown(
    dates: dates,
     theme: (BuildContext context, Widget child) {
        return Theme(
        data: ThemeData(
            buttonTheme:
                ButtonThemeData(textTheme: ButtonTextTheme.accent),
            accentColor: Colors.red,
            primaryColor: Colors.red,
        ),
        child: child,
        );
    },
    dropdownIconColor: Colors.red,
    onChanged: (DatePickerModel value) => setState(
        () {
        _selected = value.id;
        _startDate = value.startDate;
        _endDate = value.endDate;
        },
    ),
    callback: (DateTimeRange picked) => setState(
        () {
        _startDate = picked.start;
        _endDate = picked.end;
        },
    ),
)
```


## Example app

Please checkout [example](https://github.com/wendellrocha/date_range_dropdown/tree/master/example).

CHANGELOG.md
## [1.0.2]

* Fix selected item not updating dropdown value.
  
## [1.0.1]

* Fix selected item not updating dropdown value.

## [1.0.0]

* Initial release.


lib/src/date_range_dropdown/date_range_dropdown.dart
import 'package:date_range_dropdown/src/models/date_picker_model.dart';
import 'package:flutter/material.dart';

class DateRangeDropdown extends StatefulWidget {
  final List<DatePickerModel> dates;
  final Function theme;
  final DateTime firstDate;
  final DateTime lastDate;
  final String hintText;
  final String initialValue;
  final Function onChanged;
  final Function callback;
  final Color dropdownIconColor;

  const DateRangeDropdown({
    Key key,
    @required this.dates,
    this.theme,
    this.firstDate,
    this.lastDate,
    this.hintText,
    this.initialValue,
    @required this.onChanged,
    @required this.callback,
    this.dropdownIconColor,
  }) : super(key: key);
  @override
  _DateRangeDropdownState createState() => _DateRangeDropdownState();
}

class _DateRangeDropdownState extends State<DateRangeDropdown> {
  var _selected;
  var _item;

  @override
  void initState() {
    _selected = widget.initialValue != null ? widget.initialValue : null;
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: EdgeInsets.symmetric(horizontal: 10),
      decoration: BoxDecoration(
        color: Theme.of(context).dividerColor,
        borderRadius: BorderRadius.circular(8),
      ),
      child: DropdownButton(
        icon: Icon(
          Icons.arrow_drop_down,
          color: widget.dropdownIconColor != null
              ? widget.dropdownIconColor
              : Colors.blue,
        ),
        iconSize: 30,
        isExpanded: true,
        value: _selected,
        underline: SizedBox(),
        hint: Text(
          widget.hintText != null ? widget.hintText : 'Pick',
          style: TextStyle(
            fontSize: 16,
          ),
        ),
        focusColor: Theme.of(context).primaryColor,
        onChanged: (newValue) {
          FocusScope.of(context).unfocus();
          if (newValue != 'custom') {
            setState(() {
              _selected = newValue;
              _item =
                  widget.dates.firstWhere((element) => element.id == newValue);
            });
            widget.onChanged(_item);
          } else {
            datePicker(context);
          }
        },
        items: widget.dates.map<DropdownMenuItem<String>>((item) {
          return DropdownMenuItem<String>(
            child: new Text(
              item.title,
              style: TextStyle(
                fontSize: 16,
              ),
            ),
            value: item.id,
          );
        }).toList(),
      ),
    );
  }

  Future<void> datePicker(_) async {
    final DateTimeRange picked = await showDateRangePicker(
      context: _,
      firstDate: widget.firstDate != null ? widget.firstDate : DateTime(1920),
      lastDate: widget.lastDate != null ? widget.lastDate : DateTime(2100),
      builder: widget.theme != null
          ? widget.theme
          : (BuildContext context, Widget child) {
              return Theme(
                data: ThemeData(
                  buttonTheme:
                      ButtonThemeData(textTheme: ButtonTextTheme.accent),
                  accentColor: Colors.blue,
                  primaryColor: Colors.blue,
                ),
                child: child,
              );
            },
    );
    if (picked != null) {
      _selected = 'custom';
      widget.callback(picked);
    }
  }
}


example/lib/main.dart
import 'package:date_range_dropdown/date_range_dropdown.dart';
import 'package:flutter/material.dart';
import 'package:jiffy/jiffy.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Date Range Dropdrown Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Date Range Dropdrown Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<DatePickerModel> dates = List<DatePickerModel>();

  var _lastMonth = Jiffy().subtract(months: 1);
  DateTime _startDate = Jiffy().startOf(Units.DAY);
  DateTime _endDate = Jiffy().endOf(Units.DAY);
  String _selected = 'today';

  @override
  void initState() {
    dates.add(DatePickerModel.fromJson({
      "id": 'today',
      "title": "Today",
      "startDate": Jiffy().startOf(Units.DAY),
      "endDate": Jiffy().endOf(Units.DAY)
    }));

    dates.add(DatePickerModel.fromJson({
      "id": 'yesterday',
      "title": "Yesterday",
      "startDate": Jiffy().startOf(Units.DAY).subtract(Duration(days: 1)),
      "endDate": Jiffy().endOf(Units.DAY).subtract(Duration(days: 1))
    }));

    dates.add(DatePickerModel.fromJson({
      "id": 'last_7_days',
      "title": "Last 7 days",
      "startDate": Jiffy().startOf(Units.WEEK),
      "endDate": Jiffy().endOf(Units.WEEK)
    }));

    dates.add(DatePickerModel.fromJson({
      "id": 'last_30_days',
      "title": "Last 30 days",
      "startDate": Jiffy().startOf(Units.DAY).subtract(Duration(days: 30)),
      "endDate": Jiffy().endOf(Units.DAY)
    }));

    dates.add(DatePickerModel.fromJson({
      "id": 'last_month',
      "title": "Last month",
      "startDate": Jiffy(_lastMonth).startOf(Units.MONTH),
      "endDate": Jiffy(_lastMonth).endOf(Units.MONTH)
    }));

    dates.add(DatePickerModel.fromJson({
      "id": 'this_month',
      "title": "This month",
      "startDate": Jiffy().startOf(Units.MONTH),
      "endDate": Jiffy().endOf(Units.MONTH)
    }));

    dates.add(DatePickerModel.fromJson({
      "id": "custom",
      "title": "Custom",
      "startDate": DateTime.now(),
      "endDate": DateTime.now()
    }));

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Container(
          padding: EdgeInsets.all(16),
          child: Column(
            children: <Widget>[
              DateRangeDropdown(
                dates: dates,
                initialValue: _selected,
                theme: (BuildContext context, Widget child) {
                  return Theme(
                    data: ThemeData(
                      buttonTheme:
                          ButtonThemeData(textTheme: ButtonTextTheme.accent),
                      accentColor: Colors.red,
                      primaryColor: Colors.red,
                    ),
                    child: child,
                  );
                },
                dropdownIconColor: Colors.red,
                onChanged: (DatePickerModel value) => setState(
                  () {
                    _selected = value.id;
                    _startDate = value.startDate;
                    _endDate = value.endDate;
                  },
                ),
                callback: (DateTimeRange picked) => setState(
                  () {
                    _startDate = picked.start;
                    _endDate = picked.end;
                  },
                ),
              ),
              SizedBox(height: 6),
              Text('start: $_startDate, end: $_endDate, selected: $_selected')
            ],
          ),
        ),
      ),
    );
  }
}
